parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -172,3 +189,27 @@ nameChanges namePrefix (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< chil
       xs
         & Set.toList
         & fmap (name,)
+
+-- | Get a 'NameBasedDiff' from a 'TreeDiff'.
+nameBasedDiff :: TreeDiff -> NameBasedDiff
+nameBasedDiff (TreeDiff (DefinitionDiffs {termDiffs, typeDiffs} :< children)) =
+  let NameBasedDiff childrenTerms childrenTypes =
+        foldMap (nameBasedDiff . TreeDiff) children
+   in NameBasedDiff
+        { terms = foldMap nameBasedTermDiff termDiffs <> childrenTerms,
+          types = foldMap nameBasedTypeDiff typeDiffs <> childrenTypes
+        }
```

Comment:
TreeDiff just wraps a Cofree, so I believe you can factor out the recursion into a single foldMap here if you like.

```suggestion
nameBasedDiff (TreeDiff d) =
  d & foldMap \defDiff -> 
        NameBasedDiff 
        { terms = nameBasedTermDiff defDiff
        , types = nameBasedTypeDiff defDiff
        }
```

Reply:
I'm not sure I understand. The suggestion doesn't type check.

Reply:
Oh, I missed a foldMap, fixed it.

It's not really any shorter, but personally I like to factor out recursion when I can, I find it helps with readability 😄